Made the compilation file, and outlined our current strategy and why the vstack (#309)
cryptiumlabs/juvix: doc/Architecture/Compilation.org

* Compilation
** Michelson

> Michelson is an interpreted stack based language on-top of the Tezos

Reviewer: ```suggestion
Michelson is an interpreted stack based language which runs on the Tezos
```

Also, perhaps link to it?

Author: Done!

chain.

Since we are compiling to Michelson, some interesting considerations
must be made
*** Push/Enter vs Eval/Apply
Although the drawbacks of Push/Enter compared to Eval/Apply are well
known [fn:1] [fn:2], for an interpretive stack based languages, the
caveats in that statement go away.
1. _Eval/Apply is easier to compile to a normal C like language_
   - Michelson is a stack based language, this point does not apply
2. _Push/Enter spread complexity around when trying to make a fast compiler_
   - Michelson is an interpreted language, and in particular, creating

>      lambdas (closures) for over compilation is quite costly. What

Reviewer: Do you mean overevaluation?

Author: yeah you are correct, changing it now!

     would be cheaper for this language would be to simply push the
     arguments onto the stack for the new function to consume, and
     inline said function.
*** Compilation Strategy
Currently annotated Core is translated directly into Michelson

This is less than ideal for more than a few reasons, and will be

> changed at a current release.

Reviewer: Changed in a future release?

Author: Fixed and thank you

However until that time, we will document what is *currently* being
done and how to improve it

One last note is that names are assumed to not be overlapping, as this
happens higher up in the compilation process

> **** Virtual Stack

Reviewer: Separately, we should talk about how we track the (real) Michelson stack while compiling (currently in order to map names to stack positions & look up types of things)

Author: Sure will do

Currently in compiling Michelson we have what we labeled the =Virtual
Stack= (vstack), This stack could all be in the real stack or none of it could
be. What could be stored on the vstack is all the various constants,
as these would waste computation if passed around normally.

This vstack was designed so that, one could easily reason about the
stack without having to worry about “If I dup this, would it really
dup it, or just dup what is on top of the *real* stack?”

> 

Reviewer: Is there a particular case where constant propagation would work with a virtual stack but not with an earlier (pre-stack) inlining pass? Can we talk about that case (or cases)?

Author: Sure I can add a bit. But basically it boils down to maybe do some smart compilation and then trying to interpret 

```factor
: add +
```

We may compile this ahead of a time, and not be sure of constants. So if we do this with a vstack, this is fine, and we can do some smart compilation without doing extra analysis.
